socs/agents/hwp_supervisor/agent.py

```diff
@@ -0,0 +1,492 @@
+                    'ups_minutes_remaining_thresh': 45.0,
+                    'ups_output_source': 'normal'  # See UPS agent docs for choices
+                },
+                 # Subsystem action recommendations determined from state data
+                'actions': {
+                    'pmx': 'ok'  # 'ok', 'stop', or 'no_data'
+                    'grippder': 'ok'  # 'ok', 'stop', or 'no_data'
+                }}
+        """
+        pm = Pacemaker(1. / self.sleep_time)
+        test_mode = params.get('test_mode', False)
+
+        session.data = {
+            'timestamp': time.time(),
+            'monitored_sessions': {},
+            'state': {},
+            'actions': {}
+        }
+
+        kw = {'test_mode': test_mode, 'log': self.log}
+
+        while session.status in ['starting', 'running']:
+            session.data['timestamp'] = time.time()
+
+            # 1. Gather data from relevant operations
+            temp_op = get_op_data(self.hwp_lakeshore_id, 'acq', **kw)
+            enc_op = get_op_data(self.hwp_encoder_id, 'acq', **kw)
+            pmx_op = get_op_data(self.hwp_pmx_id, 'acq', **kw)
+            pid_op = get_op_data(self.hwp_pid_id, 'acq', **kw)
+            ups_op = get_op_data(self.ups_id, 'acq', **kw)
+
+            session.data['monitored_sessions'] = {
+                'temperature': temp_op,
+                'encoder': enc_op,
+                'pmx': pmx_op,
+                'pid': pid_op,
+                'ups': ups_op
+            }
+
+            # gather state info
+            state = {}
+            self._update_state_temp(temp_op, state)
+            self._update_state_ups(ups_op, state)
+
+            if enc_op['status'] == 'ok':
+                state['hwp_freq'] = enc_op['data']['approx_hwp_freq']
+            else:
+                state['hwp_freq'] = None
+
+            session.data['state'] = state
+
+            # Get actions for each hwp subsystem
+            session.data['actions'] = {
+                'pmx': self._get_pmx_action(state),
+                'gripper': self._get_gripper_action(state),
+            }
+
+            if test_mode:
+                break
+
+            pm.sleep()
+
+        return True, "Monitor process stopped"
+
+    def _stop_monitor(self, session, params):
+        session.status = 'stopping'
+        return True, 'Stopping monitor process'
+
+    @ocs_agent.param('forward', type=bool, default=True)
+    @ocs_agent.param('target_freq', type=float)
+    def spin_up(self, session, params):
+        """spin_up(forward=True, target_freq=2.0)
+
+        *Task* -- Sets the HWP to spin at a given frequency.
+
+        An example of the session data::
+
+            >>> response.session['data']
+
+                {'forward': True,
+                'commanded_freq': 2.0,
+                'pid_freq': 1.2}
+
+        Parameters
+        ------------
+        forward : bool
+            If True, will spin the HWP in the forward direction.
```

Comment:
Thanks for pointing it out.

This is not described in the so coordinate system document yet, but it is [proposed](https://docs.google.com/presentation/d/1LKq5hsVltEiDbHQZ6quGuD-I8CXO6oDZtlnc_la0bN8/edit?usp=sharing) to define the "forward hwp rotation direction" as "counter-clockwise from the sky", so I think we should define it that way.

I think the "forward" and "reverse" for PID agent is fine as it is, because PID agent is just turning on or off one of the digital outputs to change rotation direction.

Because of the slight differences of the rotation mechanism of the three SATs, the definition of "forward" and "reverse" for PID agent is not physically consistent among three SATs.
So, I think we need a config file for hwp supervisor agent.

Reply:
Can I suggest that instead of separate arguments for direction and rate, that only a single argument is given: the rate, as a _signed float_.  So `target_freq=2`. would spin at 2Hz CCW, `target_freq=-2` would spin at 2 Hz CW.

Also ... some people think that we'll only ever spin the HWP one direction.  But not I.  I want to spin it both directions, equally.  So defaulting to forward, or not forward, is not what I would expect.

So the HWP supervisor needs to be configurable, with regards to how +ve / -ve spin becomes a direction for the PID agent... that can just be command line arguments to the HWP supervisor agent.  The "config file" is the OCS site config file.  (I'm not against having a separate config file ... it's just not totally needed here.)